Allow @available(swift, ...) nested within @available(macOS, ...)

jrose-apple opened the pull request:
Previously, we treated this as an attempt to widen the availability of a member beyond its context, but it's really a different axis of availability.

Also, warn about using `@available` on extensions without an OS, which we just completely ignore right now.

rdar://problem/32632327

jrose-apple commented:
@swift-ci Please test

jrose-apple commented:
@swift-ci Please test source compatibility

Coeur commented:
What about a test for multiple attributes, like:

```
@available(macOS 10.12, *)
@available(swift 400)
extension DummyType {}
```

and

```
@available(swift 400)
@available(macOS 10.12, *)
extension DummyType {}
```

Should it warn?

jrose-apple commented:
Ah, I see no reason why the existing code wouldn't work for that, but it's a good test to have! Thanks for the suggestion.

jrose-apple committed:
Allow @available(swift, ...) nested within @available(macOS, ...)

Previously, we treated this as an attempt to widen the availability
of a member beyond its context, but it's really a different axis of
availability.

Also, warn about using @available on extensions without an OS, which
we just completely ignore right now.

rdar://problem/32632327

jrose-apple committed:
[stdlib] Remove ignored platform-agnostic '@available' from extensions

It looks like we exposed a bunch of Metal APIs to Swift 3 that we
didn't mean to; at this point it would be a source-breaking change to
hide them. Oops.

jrose-apple commented:
@swift-ci Please test

moiseev reviewed (approved):
:shipit:

jrose-apple commented:
Hm, it's the two failures that @davezarzycki was seeing earlier. Except I *did* rebase…

jrose-apple commented:
@swift-ci Please test macOS

davezarzycki commented:
Hi @jrose-apple – I may be grasping at straws here, but I'm seeing seemingly random but repeatable/inexplicable test time outs in other scenarios too. For example, @xedin made what a seemingly harmless change in #15134 that caused two simple/fast tests to start timing out on my Linux box. Do the timeouts in this pull request go away if you remove `-swift-verion 3`? That worked on my Linux box for #15134.

xedin commented:
It seems like these failures are sporadic, I saw them happen couple of times on Linux CI even before my changes went in.

jrose-apple commented:
I'm not sure that what you saw on Linux was related to the OpenCL and SceneKit tests on macOS, because it's *always* those two tests. I asked Mishal to take a look at the particular bot that's been running our tests—turns out it was one specific machine for at least my failure and one of yours—and he said he'd take it offline for now.

Linux tests have been bumping up against the timeouts for a while over the last few months. I wonder if there are one or two long-running tests there and/or if we've regressed on something, but no one's looked into it yet that I know of.

DougGregor reviewed (approved):
LGTM,
